adapters/adhese/adhese.go

```diff
@@ -0,0 +1,259 @@
+	return fmt.Sprintf("/sl%s-%s", parameters.Location, parameters.Format)
+}
+
+func extractTargetParameters(parameters openrtb_ext.ExtImpAdhese) string {
+	if parameters.Keywords == nil || len(parameters.Keywords) == 0 {
+		return ""
+	}
+	m := make(map[string][]string)
+	var targetParsed map[string]interface{}
+	json.Unmarshal(parameters.Keywords, &targetParsed)
+	for targetKey, targetRawValue := range targetParsed {
+		var targetingValues = targetRawValue.([]interface{})
+		for _, targetRawValKey := range targetingValues {
+			var targetValueParsed = targetRawValKey.(string)
+			cur, _ := m[targetKey]
+			new := cur[:]
+			m[targetKey] = append(new, targetValueParsed)
+		}
+	}
+
+	var parametersAsString = ""
+
+	for k, v := range m {
+		parametersAsString += "/" + k + strings.Join(v, ";")
+	}
+	params := strings.Split(parametersAsString, "/")
+	sort.Strings(params)
+	return strings.Join(params, "/")
+
+}
+
+func extractGdprParameter(request *openrtb.BidRequest) string {
+	if request.User != nil {
+		var extUser openrtb_ext.ExtUser
+		if err := json.Unmarshal(request.User.Ext, &extUser); err == nil {
+			return "/xt" + extUser.Consent
+		}
+	}
+	return ""
+}
+
+func extractRefererParameter(request *openrtb.BidRequest) string {
+	if request.Site != nil && request.Site.Page != "" {
+		return "/xf" + url.QueryEscape(request.Site.Page)
+	}
+	return ""
+}
+
+func (a *AdheseAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+
+	var err error
+	var imp = &request.Imp[0]
+	var bidderExt adapters.ExtImpBidder
+
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		errs = append(errs, WrapError("Request could not be parsed as ExtImpBidder due to: "+err.Error()))
+		return nil, errs
+	}
+
+	var params openrtb_ext.ExtImpAdhese
+	if err := json.Unmarshal(bidderExt.Bidder, &params); err != nil {
+		errs = append(errs, WrapError("Request could not be parsed as ExtImpAdhese due to: "+err.Error()))
+		return nil, errs
+	}
+
+	// Validate request
+	if params.Account == "" || params.Location == "" || params.Format == "" {
+		errs = append(errs, WrapError("Request is missing a required parameter (Account, Location and/or Format)"))
+		return nil, errs
+	}
+
+	// Compose url
+	endpointParams := macros.EndpointTemplateParams{Host: "ads-" + params.Account + ".adhese.com"}
+
+	host, err := macros.ResolveMacros(*&a.endpointTemplate, endpointParams)
+	if err != nil {
+		errs = append(errs, WrapError("Could not compose url from template and request account val: "+err.Error()))
+		return nil, errs
+	}
+	complete_url := fmt.Sprintf("%s%s%s%s%s",
+		host,
+		extractSlotParameter(params),
+		extractTargetParameters(params),
+		extractGdprParameter(request),
+		extractRefererParameter(request))
+
+	// If all the requests are invalid, Call to adaptor is skipped
+	if len(request.Imp) == 0 {
```

Comment:
It is better place this check before line 86 where we dereference the first element of `request.Imp`
```
86         	var imp = &request.Imp[0]
```

Reply:
Correct